### ISSUE TITLE
Cannot insert new list items with more constrained content in schema

### ISSUE THREAD

**Issue by theefer** [ProseMirror/prosemirror#443]
Editing lists works fine with a default Schema like:

``` js
const schema = new Schema({
  nodes: {
    doc: {type: Doc, content: "block+"},
    paragraph: {type: Paragraph, content: "inline<_>*", group: "block"},
    ordered_list: {type: OrderedList, content: "list_item+", group: "block"},
    bullet_list: {type: BulletList, content: "list_item+", group: "block"},
    list_item: {type: ListItem, content: "paragraph block*"},
    // ...                                          ^^^^^^
```

but when constraining the `content` allowed in a `list_item`:

``` js
    list_item: {type: ListItem, content: "paragraph (ordered_list | bullet_list)?"},
```

trying to press enter at the end of a `list_item` throws the following error:

```
transform:42 Uncaught Error: Cannot join paragraph onto ordered_list
ProseMirrorError	@	error:7
TransformError	@	transform:5
step	@	transform:42
Transform.split	@	structure:180
commands.splitBlock	@	commands:292
(anonymous function)	@	commands:28
handle	@	input:60
dispatchKey	@	input:65
handlers.keydown	@	input:125
Input.pm.content.addEventListener.e	@	input:35
```

I am able to create a new list (eg by typing `* abc`), just not create subsequent items in it.

**Comment by marijnh** [comment]
Are you binding enter to the `splitListItem` command from the distribution? If so, yes, that makes assumptions about the schema, and is only suitable for a list definition like the one in the base schema.

**Comment by theefer** [comment]
Ah good question. After tracking down the issue a little deeper, it turns out that the error above happens whether the default `splitListItem` is used or not. With the Schema I outlined, `splitListItem` actually exits early (it detects that it [can't split](https://github.com/ProseMirror/prosemirror/blob/97addcc4cea114751ce5a6e46680efc22a5f0b24/src/commands-list/index.js#L68) and returns `false`), so the error is happening in the default call to `splitBlock` which I assume is [configured by the baseKeymap](https://github.com/ProseMirror/prosemirror/blob/b672b25e22ff4e19f911090043b21e760b03603a/src/commands/index.js#L561).

Should `splitBlock` be made to cope with Enter in such a Schema? Or would you expect I should write my own `splitListItem` override that would cope with the different `list_item` content I expect?

**Comment by theefer** [comment]
The issue with the default `splitListItem` and the Schema I described above (the more constrained `list_item` content of `"paragraph (ordered_list | bullet_list)"`) seems to be with the [computed `nextType`](https://github.com/ProseMirror/prosemirror/blob/97addcc4cea114751ce5a6e46680efc22a5f0b24/src/commands-list/index.js#L66). As I understand it, when the selection is collapsed, we compute the `nextType` for the split by querying the `grandParent` (the ListItem node) for its default content type *after the current content*.

In our case, the current content is a `paragraph` and the default content type to succeed it is `ordered_list` (or `bullet_list`, but first in order prevails). However, this is not a valid `nextType` for a split (as tested by `canSplit`), presumably because as per the schema, a `list_item` must start with a `paragraph`.

I'm not sure I understand why the current `splitListItem` code chooses `nextType` based on what type would be next *if no split happened* (i.e. the next type that should appear next in the `list_item` according to the `content` definition). In the default schema (where `list_item` contains `paragraph block*`), that successor type is `block*`, the default of which resolves to `paragraph`, which happens to also be the type expected at the beginning of a `list_item`.

Wouldn't it make more sense for `nextType` to be the default type allowed as initial node in a `list_item`, since we're about to split the content and put the rest in a new `list_item`? I have tried doing that by swapping out the `nextType` with the following:

``` js
  let nextType = $to.pos == $from.end() ? grandParent.defaultContentType(0) : null
```

and it seems to work fine. All tests pass too.

It's very likely I'm missing something here, but please let me know your thoughts. Also, if my suggestion above is breaking a use case, would it be possible to add that case to the test suite so I can keep digging at a more correct solution, and learn about how ProseMirror works? :smile:

**Comment by theefer** [comment]
Fix was submitted in #450
